sdk/tools/whitelister/whitelister.cpp

```diff
@@ -0,0 +1,411 @@
+    if (attributes == 0)
+    {
+        attributes = WLA_BROKEN_ROS |
+                     WLA_BROKEN_WIN2003 |
+                     WLA_BROKEN_WINVISTA |
+                     WLA_BROKEN_WIN7 |
+                     WLA_BROKEN_WIN8 |
+                     WLA_BROKEN_WIN81 |
+                     WLA_BROKEN_WIN10;
+    }
+
+    Trace("Attributes = 0x%x\n", attributes);
+    return attributes;
+}
+
+void
+AddWhitelistEntry(vector<WL_ENTRY> &Whitelist, string& Line, string& HeadHash, const string& WhiteListFileName)
+{
+    size_t start, length;
+
+    // Format:
+    // <Hash> <original file name> <original line> "<reason>"
+    // 8400fdc0786b rostests/apitests/advapi32/LockDatabase.c 701 "Fails on WHS"
+
+    // Start with the hash, which is 12 characters
+    string oldHash = Line.substr(0, 12);
+
+    // Get the original file name
+    start = 13;
+    length = Line.substr(start).find(' ');
+    string oldFileName = Line.substr(start, length);
+    Trace("oldFileName = '%s'\n", oldFileName.c_str());
+
+    // Get the original line
+    start += length + 1;
+    length = Line.substr(start).find(' ');
+    int oldLine = stoi(Line.substr(start, length));
+
+    // Get the reason
+    start += length + 1;
+    length = Line.length() - start - 2;
+    string reason = Line.substr(start + 1, length);
+
+    // Get attributes from reason
+    uint32_t attributes = GetWhiteListAttributesFromReason(reason);
+
+    // Translate the information to how it looks like at the current head
+    string translated = git_rev_blame(oldHash, oldFileName, oldLine);
+    Trace("Translated: \"%s\"\n", translated.c_str());
+
+    // Format:
+    // // <new hash> <new file name> <new line> <original line>) <text>
+    // f2bc1f0e1118 modules/rostests/apitests/advapi32/LockServiceDatabase.c 699 701)     ok_err(ERROR_SUCCESS);
+
+    // get the new hash
+    string newHash = translated.substr(0, 12);
+
+    // Get the new file name
+    start = 13;
+    length = translated.substr(start).find(' ');
+    string newFileName = translated.substr(start, length);
+    Trace("newFileName = '%s'\n", newFileName.c_str());
+
+    // Get the new line
+    start += length + 1;
+    length = translated.substr(start).find(' ');
+    int newLine = stoi(translated.substr(start, length));
+
+    // Try to find the file in the whitelist
+    for (auto& wlEntry : Whitelist)
+    {
+        if (newFileName == wlEntry.FullFileName)
+        {
+            wlEntry.Lines.push_back(WL_LINE_ENTRY{ newLine, attributes, reason });
+            return;
+        }
+    }
+
+    // Check if the file has local changes
+    if (git_diff_check(newFileName))
+    {
+        string message = "ERROR: File with white-list entries '" + newFileName;
+        message += "' Has uncommitted changes in the working tree.\n";
+        message += "Please commit your changes first!";
+        fprintf(stderr, "%s", message.c_str());
+        throw exception();
+    }
```

Comment:
Does this mean that you have to commit before build each time you change something in rostests ?

Reply:
This is how it currently works, yes, but I might change it, since compiling the lists takes quite some time. It shouldn't be an issue though, because I expect this to basically only happen for winetest sync, as apitests should be free of this. If you are working on an apitest, which has a whitelist, just fix the test first ;-).
Anyway, I will probably change it to a pre-commit hook or something, which then checks for a revision mismatch between the sources and the compiled whitelist and asks to manually recompile the whitelists in source.

Reply:
Well, sometimes I tinker with winetests to e.g. avoid running a whole test while trying to pinpoint a bug. I'll have to live with that.

Reply:
If it uses a git-hook, you wouldn't need to commit before building, but you would need to rebuild the whitelist before committing. I can also try a quick check feature, which would only check, whether the compiled whitelist has the hash of the last commit in that folder, and if not, recompile.